### ISSUE TITLE
CarPlay map view never calls MGLMapViewDelegate method for customizing user location annotation

### ISSUE THREAD

**Issue by flashspys** [mapbox/mapbox-navigation-ios#1734]
**Mapbox Navigation SDK version:** 0.21

Hey, 
we want to display a custom user location annotation on CarPlay. At the moment we're trying this:

```
CarPlayManager.shared.delegate = self
CarPlayManager.shared.application(application, didConnectCarInterfaceController: interfaceController, to: window)
if let mapView = CarPlayManager.shared.carWindow?.rootViewController?.view as? NavigationMapView {
  mapView.delegate = CarPlayMapViewDelegate.shared
  mapView.navigationMapDelegate = CarPlayMapViewDelegate.shared
}
```
Unfortunately, the mapView does not call the `(navigation)mapView(_ mapView: MGLMapView, viewFor annotation: MGLAnnotation)` method. The same approach works on iOS. My guess is that the mapView calls the method before we are the delegate. Can you reproduce this?

**Comment by akitchen** [comment]
Hello,

I'm afraid there is not quite enough context in your ticket to reproduce. Could you please post an entire method override implementation so we can better understand how your code is structured? It does sound like a problem with call order, but is difficult to diagnose without more context.

Thank you

**Comment by flashspys** [comment]
Hello,

Assume we have a simple CarPlay app. Everything is working fine, the map shows up and also centers to the current user location. Now I want to change the userLocationAnnotationView. To do this we want, in general, get the `mapView(_ mapView: MGLMapView, viewFor annotation: MGLAnnotation) -> MGLAnnotationView` callback, wait for the `MGLUserLocationAnnotationView` and return our own view.
To get the delegate call we have to set the mapView's delegate. We do this here:

```
func application(_ application: UIApplication, didConnectCarInterfaceController interfaceController: CPInterfaceController, to window: CPWindow) {
  CarPlayManager.shared.delegate = self
  CarPlayManager.shared.application(application, didConnectCarInterfaceController: interfaceController, to: window)
  if let mapView = CarPlayManager.shared.carWindow?.rootViewController?.view as? NavigationMapView {
    mapView.delegate = CarPlayMapViewDelegate.shared
    mapView.navigationMapDelegate = CarPlayMapViewDelegate.shared
}
```

The delegate is very simple:
```
class MapViewDelegate: NSObject, NavigationMapViewDelegate, MGLMapViewDelegate {
    
    static var shared = MapViewDelegate()
    
    func navigationMapView(_ mapView: MGLMapView, viewFor annotation: MGLAnnotation) -> MGLAnnotationView? {
// breakpoint here never hits
        if annotation is MGLUserLocationAnnotationView {
            print(" \\o/ 🎉 ")
        }
        return nil
    }
    
    func mapView(_ mapView: MGLMapView, viewFor annotation: MGLAnnotation) -> MGLAnnotationView? {
// breakpoint here never hits
        if annotation is MGLUserLocationAnnotationView {
            print(" \\o/ 🎉 ")
        }
        return nil
    }
}
```

The delegate methods here are not called. That is our problem. And for this reason we can't set our custom user location view.

I hope this helped a bit more. Thank you for your support.

**Comment by flashspys** [comment]
@akitchen hey, do you have any feedback on this?

**Comment by akitchen** [comment]
Hello, sorry for the delay here.

This will be addressed by https://github.com/mapbox/mapbox-navigation-ios/pull/1601 which is still in progress and has been held up by other work. In general, many of the MGLMapViewDelegate methods are being consumed by the NavigationMapView, and we are working to expose them through NavigationMapViewDelegate

**Comment by akitchen** [comment]
#1601 was released in v0.24.0 - we'd be curious to hear if it helps you implement this. Thanks!

**Comment by flashspys** [comment]
@akitchen I just updated to 0.24.0 and tried it out but unfortunately its still not working:
In `didConnectCarInterfaceController…` we do:

```
if let mapView = CarPlay.shared.manager.carWindow?.rootViewController?.view as? NavigationMapView {
            mapView.delegate = MapViewDelegate.shared
            mapView.navigationMapDelegate = MapViewDelegate.shared
            mapView.styleURL = TurnByTurn.Config.styleURL
}
```
Our `MapViewDelegate` implements at least `navigationMapView(_ mapView: MGLMapView, viewFor annotation: MGLAnnotation)` and `mapView(_ mapView: MGLMapView, viewFor annotation: MGLAnnotation)`. None of these methods are hitted by a breakpoint.

**Comment by 1ec5** [comment]
Based on the code snippets above, you’re hooking up the NavigationMapView’s delegate correctly. You should see most of the MGLMapViewDelegate methods get called, including `mapView(_:didFinishLoading:)`. Unfortunately, MGLMapView doesn’t call `MGLMapViewDelegate.mapView(_:viewFor:)`.

As of c1c30ed07de5fe41f42baefc5850f52759f5ae41, `CarPlayMapViewController.viewDidLoad()` forces MGLMapView’s `followWithCourse` user tracking mode in order to achieve the visual appearance of a puck, as opposed to the dot that appears by default. (This isn’t quite the same puck that appears in turn-by-turn navigation.)

https://github.com/mapbox/mapbox-navigation-ios/blob/cc255c4d8b66bfab1150e321e696d54cf4debe76/MapboxNavigation/CarPlayMapViewController.swift#L103

This happens really early, within `CarPlayManager.application(_:didConnectCarInterfaceController:to:)`, before the application has gotten a chance to set the delegate:

```
#0	0x000000010f5d6542 in ::-[MGLMapView setShowsUserLocation:](BOOL) at /path/to/mapbox-gl-native-3/platform/ios/src/MGLMapView.mm:5093
#1	0x000000010de25554 in NavigationMapView.showsUserLocation.setter at /path/to/mapbox-navigation-ios/MapboxNavigation/NavigationMapView.swift:128
#2	0x000000010de25133 in @objc NavigationMapView.showsUserLocation.setter ()
#3	0x000000010f5d7e88 in ::-[MGLMapView setUserTrackingMode:animated:](MGLUserTrackingMode, BOOL) at /path/to/mapbox-gl-native-3/platform/ios/src/MGLMapView.mm:5206
#4	0x000000010de5073b in CarPlayMapViewController.viewDidLoad() at /path/to/mapbox-navigation-ios/MapboxNavigation/CarPlayMapViewController.swift:103
#5	0x000000010de50784 in @objc CarPlayMapViewController.viewDidLoad() ()
#6	0x000000011dfd84e1 in -[UIViewController loadViewIfRequired] ()
#7	0x000000011dfd8940 in -[UIViewController view] ()
#8	0x000000011e62fc53 in -[UIWindow addRootViewControllerViewIfPossible] ()
#9	0x000000011e630c2c in -[UIWindow setRootViewController:] ()
#10	0x000000010decf627 in CarPlayManager.application(_:didConnectCarInterfaceController:to:) at /path/to/mapbox-navigation-ios/MapboxNavigation/CarPlayManager.swift:175
```

**Comment by 1ec5** [comment]
This call to `MGLMapView.setUserTrackingMode(_:animated:)` is redundant to this call that is guaranteed to occur after the delegate is set:

https://github.com/mapbox/mapbox-navigation-ios/blob/cc255c4d8b66bfab1150e321e696d54cf4debe76/MapboxNavigation/CarPlayManager.swift#L291

**Comment by flashspys** [comment]
When does this behavior get fixed? Is there a possible workaround?

**Comment by 1ec5** [comment]
A fix is ready in #1852.
